Reg compacted datasets in Hive

sahilTakiar opened the pull request:
* Initial PR to discuss the approach for adding Hive-Reg for datasets produced by `MRCompactor`
* Registration is done in a `AfterCompactionAction` once compaction has completed
* The `AfterCompactionAction`s are configurable and are invoked in `JobRunnerExecutor.afterExecute`

@zliu41 can you review?

zliu41 commented:
This creates a HiveMetaStoreClient connection for each dataset. Can you create a HiveRegister object in MRCompactor and use it to register all datasets?

sahilTakiar commented:
@zliu41 addressed comments and modified approach slightly

* Instead of making a listener specific to `MRCompactor` I made a generic `CompactorListener`
* Added factory interfaces and implementations for creating `Compactor`s and `CompactorListener`s
* Re-factored `AzkabanCompactionJobLauncher` a bit to account for changes
* Main logic for this feature is in the class `HiveRegistrationCompactorListener`

zliu41 commented:
We need a solution for retrying failed Hive registration, to avoid a dataset being published but not registered with Hive. Ideally this should leverage exactly-once, but since it likely won't be ready by Q1, a temp solution is needed, e.g., write failed dataset paths to an HDFS location which can be picked up by the next run to retry.

zliu41 commented on gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java:
```diff
@@ -788,6 +792,13 @@ protected void afterExecute(Runnable r, Throwable t) {
             // Set the dataset status to COMPACTION_COMPLETE if compaction is successful.
             jobRunner.getDataset().setState(COMPACTION_COMPLETE);
           }
+          if (MRCompactor.this.compactorListener.isPresent()) {
```
This seems to allow only one listener. Can you make it possible to have multiple listeners?

sahilTakiar replied:
Updated, it can now run multiple listeners serially.

zliu41 replied:
What is the change? seems there is still only one listener here. I thought `MRCompactor` should contain a collection of listeners.

sahilTakiar replied:
There is a new class `SerialCompactorListener` that takes in a `List` of `CompactorListener`s and executes them one by one. This class is used by `ReflectionCompactorListenerFactory`.

sahilTakiar commented:
Yes, I will focus on the retry logic in the next PR as it may require some additional design work.

zliu41 commented:
LGTM, please add copyright block to all new files
